### PR TITLE
LPD-17452 Allow admins to save the definition for list as visualization mode

### PR DESCRIPTION
**Pull request by markocikos**
### References

- Parent story: [LPD-6624 Allow admins to save the definition for list as visualization mode](https://issues.liferay.com/browse/LPD-6624)

### What is the goal of this PR?

In this PR we are:
- creating persistence for list visualization mode.
- implementing saving and editing of list sections (list elements). This included some refactoring of UI implementation of list implementation, in naming and internal structures @p2kmgcl  
- doing some minor refactoring with utils locations.

See technical notes in inline comments.

### What does it look like?

https://github.com/liferay-frontend/liferay-portal/assets/5435511/dd299ce9-7e35-4e20-a328-a88b085f9934

### CODE REVIEW

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/portlet/FDSViewsPortlet.java
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/portlet/FDSViewsPortlet.java:500]: The main idea for the new object entry `FDSListSection` is to represent building blocks of list visualization mode. In that way, it is equivalent to `FDSField`, the building block of table visualization mode. `FDSField` is really unfortunately named, it should be `FDSTableField` or `FDSTableColumn`. `FDSListSection` and `FDSField` are both entities that are 1:1 linked with `fieldName`, the main point of wiring to data definition (OpenAPI) and endpoints returned data. They are visual representation of schema property of a given `fieldName`. 

I added `rendererName` as optional attribute. We will not use it any time soon, but I am completely sure we will use it eventually. It serves to highlight why I chose this kind of data model.

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/components/FieldSelectModalContent.tsx
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/components/FieldSelectModalContent.tsx:156]: There is a high amount of code duplication with this component and `AddFieldsModalContent`. That component served as a starting point, but I took out saving logic. The goal was to make this component only for selection, reusable, with saving itself done in parent contexts.

I left in field specifics, as fields are the only hierarchical select in DSM scope. This could be made more abstract, something like a `TreeviewSelectModalContent`, or even `TreeviewSelect`, but that would require more extensive API and taking out many DSM specifics. That may be a future task.

The intent is to migrate `AddFieldsModalContent` to use `FieldSelectModalContent`. This can be in a followup, this PR is already big enough.

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/components/FieldSelectModalContent.tsx
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/components/FieldSelectModalContent.tsx:294]: The modal UI does not match mockups:
- there should be no checkboxes
- we should only allow selection of leaf items

We can do this in a [separate story.](https://liferay.atlassian.net/browse/LPD-6702)

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/visualization_modes/list/List.tsx
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/visualization_modes/list/List.tsx:41]: This was the largest refactor I needed to do. The `Record` structure was not sufficient, I needed several attributes per building block, contained in `IListSection`.

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/visualization_modes/list/List.tsx
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/visualization_modes/list/List.tsx:35]: - `IFDSListSection` represents the stored entity.
- `IListSection` represents the visual representation of the stored entity.

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by markocikos** [comment]
No unique failures, based on https://github.com/liferay-frontend/liferay-portal/pull/3990#issuecomment-1943947360. Waiting on peer review now.

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/visualization_modes/list/List.tsx
Comment (p2kmgcl) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/visualization_modes/list/List.tsx:199]: So now it is onSelect instead of onSave

**Comment by p2kmgcl** [comment]
Hey Marko! I think this is good to go :smile:

I had some comments on the typings (things I think that can be simplified), but I want this to be merged, and all the comments I had were a bit picky :laughing: 

I agree that we should have a look at all naming, because it is getting quite confusing and hard to follow (some items with prefix, some without it, sections, views, data...).

**Review by p2kmgcl — APPROVED** [review]

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/visualization_modes/list/List.tsx
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/visualization_modes/list/List.tsx:199]: Yep. The `Save` button no longer does the saving, only returns selected field, so I had no choice :smile:

**Comment by markocikos** [comment]
> I had some comments on the typings (things I think that can be simplified)

@p2kmgcl Feel free to go wild and make those changes in your [next PR](https://github.com/liferay-frontend/liferay-portal/pull/3988). I'll wait for that PR to be done, or close to done, before I start with cards persistence. Unless something drastically changes, it'll be the same, `FDSCardsSection`.

**Comment by juanjofgliferay** [comment]
Amazing :tada: . Great job!

**Comment by markocikos** [comment]
> we should have a look at all naming, because it is getting quite confusing and hard to follow (some items with prefix, some without it, sections, views, data...).

@p2kmgcl There are a few notes on naming changes that I forgot to mention, that are not obvious:

1. A `field` in DSM code is the entity that you choose in modal. Its origin is OpenAPI, where it is called schema property. The most important part of it is `fieldName`, as that wires to data returned from endpoints, item property. That's part of the reason why I renamed `fieldValues`, to avoid `field` referring to something else.
2. `FDS` prefix always represents the a entity, `FDSEntry`, `FDSView`, `FDSListSection` etc. We are very consistent about this. That's why, in this PR, we have `getFDSListSections`, `IFDSListSection` and `fdsListSection` variable. On the other hand, structure that we fill in frontend can be anything, it depends on code organization and UI. In the PR I chose `IListSection` / `listSection`, but it can be anything, that only partially contains `FDSListSection` entry.

**Pull request closed without merging**
